### ISSUE TITLE
Service worker does not update response headers

### ISSUE THREAD

**Issue by rahimalizada** [angular/angular#32770]
# 🐞 bug report

### Affected Package
@angular/pwa
@angular/service-worker

### Is this a regression?
No

### Description
Response headers are not updated when service worker updates assets.

## 🔬 Minimal Reproduction
Example with CSP header:
1. Add restricting Content Security Policy header on production server, like `img-src 'self'`
2. Add a violation to the restriction in you app (link to an image on different host)
3. Open angular PWA enabled application and ensure that CSP blocks that violation (in production)
4. Change CSP policy on the server to allow image requests to that host `img-src 'self' https://imagehost.com`
5. When the application is reloaded, service worker does not update neither assets nor headers for assets
6. When application is rebuilt and reloaded, service worker updates assets, but headers are not updated.

## 🌍  Your Environment

**Angular Version:**
<pre><code>

     _                      _                 ____ _     ___ 
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | | 
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | | 
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 8.3.4
Node: 12.9.1
OS: win32 x64
Angular: 8.2.6
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router, service-worker

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.803.4
@angular-devkit/build-angular      0.803.4
@angular-devkit/build-ng-packagr   0.803.4
@angular-devkit/build-optimizer    0.803.4
@angular-devkit/build-webpack      0.803.4
@angular-devkit/core               8.3.4
@angular-devkit/schematics         8.3.4
@angular/cdk                       8.2.0
@angular/cli                       8.3.4
@angular/material                  8.2.0
@angular/pwa                       0.803.4
@ngtools/webpack                   8.3.4
@schematics/angular                8.3.4
@schematics/update                 0.803.4
ng-packagr                         5.5.0
rxjs                               6.5.3
typescript                         3.5.3
webpack                            4.39.2
</code></pre>

**Anything else relevant?**
Tested in Chrome 76.0.3809.132 and Firefox 69.0.1

**Comment by gkalpak** [comment]
Thx for reporting this issue. Can you, please, add some more info? E.g.:
- What does your `ngsw-config.json` file look like?
- Which assets did you expect/need to get updated?

<sub>
From what I understand, what you describe is intended behavior. In order to conserve network bandwidth, the ServiceWorker will not re-download resources if it can find them in a previous cache and their content hasn't changed. Even of the headers might be different, the SW will never know, because it will not make the request to the server.
</sub>

**Comment by rahimalizada** [comment]
> What does your ngsw-config.json file look like?

ngsw-config.json is default one generated by @angular/pwa:
<pre><code>{
  "$schema": "../../node_modules/@angular/service-worker/config/schema.json",
  "index": "/index.html",
  "assetGroups": [
    {
      "name": "app",
      "installMode": "prefetch",
      "resources": {
        "files": [
          "/favicon.ico",
          "/index.html",
          "/manifest.webmanifest",
          "/*.css",
          "/*.js"
        ]
      }
    }, {
      "name": "assets",
      "installMode": "lazy",
      "updateMode": "prefetch",
      "resources": {
        "files": [
          "/assets/**",
          "/*.(eot|svg|cur|jpg|png|webp|gif|otf|ttf|woff|woff2|ani)"
        ]
      }
    }
  ]
} </code></pre>

> Which assets did you expect/need to get updated?

Assets are updated correctly.

The ServiceWorker caches assets and response headers for those assets correctly, but response headers are not updated when ServiceWorker re-downloads changed assets. So, the app gets updated, but is stuck with invalid headers (in my case CSP headers)

**Comment by gkalpak** [comment]
> The ServiceWorker caches assets and response headers for those assets correctly, but response headers are not updated when ServiceWorker re-downloads changed assets.

That would be very unexpected. I would understand if the SW did not download unchanged assets and instead copied them over from the old cache. But are you saying that the SW downloads changed assets but keep the old headers? I can't see how this could happen 😕

**Comment by cbrandt713** [comment]
I can also confirm this has been happening to me. It's a very similar case, where the Content-Security Policy headers are very out of date, in my case, Content-Security headers are still being cached from roughly a month ago.

Unfortunately I can't give a simple reproduction of the process, and the web app is closed-source.

When bypassing the service worker for the network in Chrome DevTools, the proper headers are used. When the service worker is toggled back on, the old headers are used and valid sources are blocked again. Hard clearing cache or forcibly deleting and reactivating the service worker fix this issue, however I can't expect my end users to do that. So the old headers, it would seem, are somehow being cached despite a new version of CSP being available.

Let me know if I can provide any more information that might be useful to resolve this.

**Comment by gkalpak** [comment]
Thx for chiming in, @cbrandt713. Can you confirm that the content of affected assets has not changed between versions and that the SW is not downloading the assets during an update?

**Comment by cbrandt713** [comment]
> Thx for chiming in, @cbrandt713. Can you confirm that the content of affected assets has not changed between versions and that the SW is not downloading the assets during an update?

As in the author's case. The affected has is changing between versions. It is the /index.html file, which is correctly being updated with the new version of the app. The headers coming in with the new csp are also accurately reflected in the devtools, but they are not actually being used by the browser.

**Comment by gkalpak** [comment]
It is probably me, but I am still not 100% clear on what is happening.

Focusing on `index.html` and an asset (say a `foo.png` image), which of the following are true:

a. The content of `index.html` has changed between versions.
b. The correct, updated content for `index.html` is served by the SW after the update.
c. `index.html` is served by the SW with incorrect CSP headers after the update.
d. The content of `foo.png` has changed between versions.
e. The correct, updated content for `foo.png` is served by the SW after the update.
f. `foo.png` is served by the SW with incorrect CSP headers after the update.

**Comment by cbrandt713** [comment]
In my case, A and B are true.
Everything appears to be working at surface level. The new index.html is served properly and cached. But when the browser throws an error for violation of CSP, it shows the value as the OLD CSP value, which hints that the SW somehow cached the old value for the CSP, even though the new updated index.html file now has a new CSP and it shows properly in DevTools.
It almost seems like a browser bug, but I managed to reproduce it on Chrome, Firefox and Edge, and disabling the service worker fixes the issue on all of them. So while I don't really understand how it happens, it seems like it has to be the service worker.

**Comment by gkalpak** [comment]
Thx for the extra info, @cbrandt713. I cannot understand how what you describe (if I understand it correctly) could happen (i.e. that the SW downloads the new `index.html`, but somehow keeps the old headers). I'm afraid we can't do much without a reproduction 😞

**Comment by cbrandt713** [comment]
Upon some further investigating I'm thinking the issue is caused by the issue in my debug log, and perhaps OP is having a similar issue. The debug log is claiming that it keeps getting the wrong hash for the index.html file, which might be a version of the file with the wrong CSP. Here is my debug log if it proves useful.

> NGSW Debug Info:

Driver state: EXISTING_CLIENTS_ONLY (Degraded due to failed initialization: Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected 0ea3522de6542bd915ec5cce1625163e00abdec1, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting)
Error: Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected 0ea3522de6542bd915ec5cce1625163e00abdec1, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting)
    at PrefetchAssetGroup.<anonymous> (https://betamyapps.classlink.com/ngsw-worker.js:699:35)
    at Generator.next (<anonymous>)
    at fulfilled (https://betamyapps.classlink.com/ngsw-worker.js:320:62))
Latest manifest hash: aea508f6dfc26921c0e3008818e85336618be276
Last update check: 1m38s916u

=== Version aea508f6dfc26921c0e3008818e85336618be276 ===

Clients: 

=== Idle Task Queue ===
Last update tick: 1m43s940u
Last update run: 1m38s938u
Task queue:


Debug log:

[9m14s685u] Error(Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected 0ea3522de6542bd915ec5cce1625163e00abdec1, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting), Error: Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected 0ea3522de6542bd915ec5cce1625163e00abdec1, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting)
    at PrefetchAssetGroup.<anonymous> (https://betamyapps.classlink.com/ngsw-worker.js:699:35)
    at Generator.next (<anonymous>)
    at fulfilled (https://betamyapps.classlink.com/ngsw-worker.js:320:62)) Error occurred while updating to manifest d0f47db0324cb88e6cc9250cebf779164256754b
[9m9s62u] Error(Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected 0ea3522de6542bd915ec5cce1625163e00abdec1, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting), Error: Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected 0ea3522de6542bd915ec5cce1625163e00abdec1, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting)
    at PrefetchAssetGroup.<anonymous> (https://betamyapps.classlink.com/ngsw-worker.js:699:35)
    at Generator.next (<anonymous>)
    at fulfilled (https://betamyapps.classlink.com/ngsw-worker.js:320:62)) Error occurred while updating to manifest d0f47db0324cb88e6cc9250cebf779164256754b
[9m3s736u] Error(Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected 0ea3522de6542bd915ec5cce1625163e00abdec1, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting), Error: Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected 0ea3522de6542bd915ec5cce1625163e00abdec1, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting)
    at PrefetchAssetGroup.<anonymous> (https://betamyapps.classlink.com/ngsw-worker.js:699:35)
    at Generator.next (<anonymous>)
    at fulfilled (https://betamyapps.classlink.com/ngsw-worker.js:320:62)) Error occurred while updating to manifest d0f47db0324cb88e6cc9250cebf779164256754b
[9m3s696u] Error(Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected d4fbd2c46f2b8b26c16b12832199caba9b5d79ed, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting), Error: Hash mismatch (cacheBustedFetchFromNetwork): https://betamyapps.classlink.com/index.html: expected d4fbd2c46f2b8b26c16b12832199caba9b5d79ed, got 605d0edc77f991dd54531fb49546ec56cada23a7 (after cache busting)
    at PrefetchAssetGroup.<anonymous> (https://betamyapps.classlink.com/ngsw-worker.js:699:35)
    at Generator.next (<anonymous>)
    at fulfilled (https://betamyapps.classlink.com/ngsw-worker.js:320:62)) initializeFully for aea508f6dfc26921c0e3008818e85336618be276

**Comment by gkalpak** [comment]
Aha! That is a totally different issue then. It is expected that the SW does not install the update (with the updated CSP headers), if it gets a hash-mismatch.

You need to figure out what the difference is between the built file (based on which the expected hash is computed) and the file that is actually served from the server. Often it is the server or some proxy (or even an FTP client) normalizing line-breaks or whitespace.

**Comment by cbrandt713** [comment]
Thank you, could Git normalizing CRLF to LF be the culprit potentially?

**Comment by rahimalizada** [comment]
Reproduction on firebase hosting:

```
ng new swtest
cd swtest 
echo '<h1>text</h1><img src="https://picsum.photos/800/600">' > src/app/app.component.html
ng add @angular/pwa
```
When served locally `ng serve ` the image is displayed under the text.

`firebase init ... `

firebase.json: 
```json
{
  "hosting": {
    "public": "dist/swtest",
    "headers": [
      {
        "source": "**",
        "headers": [
          {
            "key": "Content-Security-Policy",
            "value": "default-src 'self'; img-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'"
          }
        ]
      }
    ],
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ]
  }
}
```

```
ng build --prod
firebase deploy
```

Console log in chrome: `Chrome: Refused to load the image 'https://picsum.photos/800/600' because it violates the following Content Security Policy directive: "img-src 'self'".` 
In ff: `Content Security Policy: The page’s settings blocked the loading of a resource at https://picsum.photos/800/600 (“img-src”).`

Modified firebase.json adding https://picsum.photos to allowed image source list and redeployed to firebase:
```
... 
  "value": "default-src 'self'; img-src 'self' 'https://picsum.photos'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'"
...
```

Obviously after refresh the service worker did not update neither files nor headers since files were not updated.

Modified content in app component, rebuilt and redeployed:
```
echo '<h1>text2</h1><img src="https://picsum.photos/800/600">' > src/app/app.component.html
ng build --prod
firebase deploy
```
On refresh SW loads updated files and headers, but the image is still blocked by CSP policy. On subsequent reloads the index page is served with proper headers, but SW is somehow still using initial headers. 

[Video 1](https://youtu.be/fESjMiY4eOs)
[Video 2](https://youtu.be/VwmLiwIbhjU)

**Comment by gkalpak** [comment]
> could Git normalizing CRLF to LF be the culprit potentially? Or the server gzipping the contents?

@cbrandt713: Gzipping doesn't affect the content (so, no) and git normalizing could only be the culprit if you build the app before normalizing and then serve the normalized content from the server (so it depends on your setup - but it's possible).

**Comment by gkalpak** [comment]
@rahimalizada: Thx for the detailed reproduction instructions and the videos 💯

It seems that you are setting an incorrect value for the `img-src` directive value. This is also indicated by the error in your console: 

> The source list for Content Security Policy directive 'img-src' contains an invalid source: ''https://picsum.photos''. It will ne ignored.

Basically, values like `self` are special and should be wrapped in quotes (`'`), but you **should not** wrap other sources (such as `https://picsum.photos`) in quotes. This causes the browser to ignore that sources, which means the `img-src` directive is essentially equivalent to `img-src: 'self'` again.

**Comment by rahimalizada** [comment]
> It seems that you are setting an incorrect value for the `img-src` directive value

@gkalpak  yes, thank you for pointing that out. The issue also persists with properly formed CSP header. Additionally SW gives 504 Gateway timeout and only hard reset fixes the issue.  

[With restricting CSP](https://youtu.be/Ig55DWnJjJY)
[With allowing CSP](https://youtu.be/JxgTkVyxr-8)
[Hard reset](https://youtu.be/liogMn-SSk4)

**Comment by gkalpak** [comment]
Oh, I see. This isa different issue (which affects ServiceWorkers+CSP in general - not specific to Angular). First of all, the "hard reset" is expected to work, because it basically bypasses the SW. In the "with allowing CSP" video, you can notice that the error message is a little different than before:

> Refused to connect to 'https://picsum.photos/800/600' because it violates the following Content Security Policy directive: "default-src: 'self'". Note that 'connect-src' was not explicitly set, so 'default-src' is used as a fallback.

Notice that is mentions `default-src` is violated and not `img-src`. This is related to how browsers handle requests from SW (in the presense of a CSP). You can read more about the problem (and how to fix it) [here][1].

##
This is essentially a duplicate of #31330 (see there for some more resources).
I am going to close this issue, because everything seems to be working as expected, but feel free to continue the discussion below.

[1]: https://qubyte.codes/blog/content-security-policy-and-service-workers
